db/hybrid_logical_vector.go

```diff
@@ -177,7 +193,48 @@ func (hlv *HybridLogicalVector) GetVersion(sourceID string) uint64 {
 	if mvEntry := hlv.MergeVersions[sourceID]; mvEntry > latestVersion {
 		latestVersion = mvEntry
 	}
-	return latestVersion
+	// if we have 0 cas value, there is no entry for this source ID in the HLV
+	if latestVersion == 0 {
+		return latestVersion, false
+	}
+	return latestVersion, true
+}
+
+// AddNewerVersions will take a hlv and add any newer source/version pairs found across CV and PV found in the other HLV taken as parameter
+// when both HLV
+func (hlv *HybridLogicalVector) AddNewerVersions(otherVector HybridLogicalVector) error {
+
+	// create current version for incoming vector and attempt to add it to the local HLV, AddVersion will handle if attempting to add older
+	// version than local HLVs CV pair
+	otherVectorCV := CurrentVersionVector{SourceID: otherVector.SourceID, VersionCAS: otherVector.Version}
+	err := hlv.AddVersion(otherVectorCV)
+	if err != nil {
+		return err
+	}
+
+	if otherVector.PreviousVersions != nil || len(otherVector.PreviousVersions) != 0 {
+		// Iterate through incoming vector previous versions, update with the version from other vector
+		// for source if the local version for that source is lower
+		for i, v := range otherVector.PreviousVersions {
+			if hlv.PreviousVersions[i] == 0 || hlv.PreviousVersions[i] < v {
+				hlv.setPreviousVersion(i, v)
+			}
+		}
+	}
+	// if current source exists in PV, delete it.
+	if _, ok := hlv.PreviousVersions[hlv.SourceID]; ok {
+		delete(hlv.PreviousVersions, hlv.SourceID)
+	}
+	return nil
+}
+
+func (hlv *HybridLogicalVector) SetMergeVectors(otherVector HybridLogicalVector) {
```

Comment:
I'd prefer to omit MergeVersion handling from this PR - I don't think there's enough test coverage to verify we're doing it properly (e.g. we need handling for existing MVs, we need pruning of MV on subsequent updated).